source/common/common/callback_impl.h

```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Upstream {
+
+/**
+ * Utility class for managing callbacks.
+ */
+template <class Callback, typename... CallbackArgs> class CallbackManager {
+public:
+  /**
+   * Add a callback.
+   * @param callback supplies the callback to add.
+   * @return const CallbackHandle* a handle that can be used to remove the callback.
+   */
+  CallbackHandle* add(Callback callback) {
+    callbacks_.emplace_back(*this, callback);
+    return &callbacks_.back();
+  }
+
+  /**
+   * Run all registered callbacks.
+   * @param args supplies the callback arguments.
+   */
+  void runCallbacks(CallbackArgs... args) {
+    for (const auto& cb : callbacks_) {
+      cb.cb_(args...);
```

Comment:
If you add tests for this, a good one would be to validate that remove yourself from the list in the context of a callback is safe, which it should be, as the entry isn't used outside that call stack.